### ISSUE TITLE
Upgrading from tokio-rustls 0.9.1 to 0.9.2 causes connections to stop processing unexpectedly

### ISSUE THREAD

**Issue by khuey** [quininer/tokio-rustls#37]
Unfortunately my code is not open source, the versions on crates.io do not correspond to the version in this repository, and this crate doesn't appear to print any logging even at the trace level, so I can't report anything more useful than that.

**Comment by quininer** [comment]
Can you describe it in detail or provide a reproducible step?

**Comment by quininer** [comment]
Since there is no detail, I can only make some guesses.

In 0.9.1, read inside is blocked by write, which causes some unexpected situations. see https://github.com/quininer/tokio-rustls/issues/32

After 0.9.2, read is no longer blocked by write, so you may need to write `write.and_then(flush).and_then(read)` instead of `write.and_then(read)`.

**Comment by quininer** [comment]
I realized that this is more like a behavior change than a bugfix, it should be a new version. :(

**Comment by jerryz920** [comment]
something related to changes in #32?

**Comment by haata** [comment]
I'm having the same sort of issue between 0.9.1 and anything afterwards (0.9.2+). My use case is a bit more complicated as I believe it's actually the capnproto-futures that has the problem but I haven't been able to locate yet where the write is taking place.

Here's an example:
https://github.com/haata/capntls/tree/tokio-rustls-0.9.1
```bash
cargo run server localhost:12345 &
cargo run client localhost:12345
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
     Running `target/debug/capntls client 'localhost:12345'`
tester@testserver.com:hello
tester@testserver.com:helloa
```
vs.
https://github.com/haata/capntls/tree/tokio-rustls-0.9.2
```bash
cargo run server localhost:12345 &
cargo run client localhost:12345
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/capntls client 'localhost:12345'`
tester@testserver.com:hello
^C
```

Normally I'd just start using 0.9.1 and move on however I have some other project dependencies that require a newer version of ring so I also need to use a more recent version of tokio-rustls.

**Comment by quininer** [comment]
Have you tried 0.9.3? It fixes a bug about flush. https://github.com/quininer/tokio-rustls/commit/fe8fafecee85191efb73aa7edaed7021a5c8beff

**Comment by haata** [comment]
Hmmm, 0.9.3 doesn't help at all, there must be something else that changed from 0.9.1 to 0.9.2...
(thanks for pointing that out!)

https://github.com/haata/capntls/tree/tokio-rustls-0.9.3
```bash
cargo run server localhost:12345 &
cargo run client localhost:12345
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/capntls client 'localhost:12345'`
tester@testserver.com:hello
^C
```

Something odd that I did notice when using wireshark (I have SSLKEYLOGFILE setup) is that the data is sent from the server but the packet containing the response ends up receiving a TCP Retransmission because the client didn't ACK the response. The TCP retransmission is ACK'd but it has a bunch more TCP options set.
I'll attach the trace shortly.

Server response packet, looks about what it should be (I'll need to run this again with 0.9.1).
![server_response](https://user-images.githubusercontent.com/23465/65572185-20919180-df1c-11e9-87e9-ce27141c3b4e.png)

TCP Retransmission
![retransmission](https://user-images.githubusercontent.com/23465/65572195-29826300-df1c-11e9-8391-a2c452c1bac9.png)

ACK of retransmission
![final_ack](https://user-images.githubusercontent.com/23465/65572208-30a97100-df1c-11e9-9055-3e6776ca09ab.png)

Here is the full trace as well. Packet 41 will show my@email.com:hello2 in the decoded form (was run on a different computer, was using hello2 instead of helloa, otherwise the code is the same).
[mykeys.txt](https://github.com/quininer/tokio-rustls/files/3650690/mykeys.txt)
[trace1.pcapng.gz](https://github.com/quininer/tokio-rustls/files/3650695/trace1.pcapng.gz)

**Comment by quininer** [comment]
Ah, this is my mistake. mio uses edge trigger, so premature read operations cause event loss.

I think this [patch](https://github.com/quininer/tokio-rustls/commit/61d005fb74f54cb575171f383baebabaa8cf6282) can solve problem and I will release a new version later.

**Comment by quininer** [comment]
I believe this problem has been resolved. if not, please reopen it or open a new issue.

**Comment by khuey** [comment]
I'm not seeing any issues on 0.10.
